multi/grlib-multi/spacewire.c

```diff
@@ -0,0 +1,829 @@
+		if (dev->lastTxDesc == SPW_TX_DESC_CNT - 1) {
+			/* Wrap around */
+			desc->ctrl |= TX_DESC_WR;
+		}
+		desc->packetLen = packet.dataLen;
+
+		uint8_t hdrLen = packet.flags & TX_DESC_HDR_LEN;
+		volatile void *txBuff = dev->txBuff[dev->lastTxDesc];
+
+		memcpy((char *)txBuff, packet.hdr, hdrLen);
+		memcpy((char *)txBuff + hdrLen, packet.data, packet.dataLen);
+
+		desc->hdrAddr = (uint8_t *)va2pa((void *)txBuff);
+		desc->dataAddr = (uint8_t *)va2pa((char *)txBuff + hdrLen);
+
+		/* Everything is set up, enable descriptor */
+		desc->ctrl |= TX_DESC_EN;
+
+		dev->txWaited[dev->lastTxDesc] = false;
+
+		/* Start transmission */
+		dev->vbase[DMA_CTRL] |= DMA_CTRL_TE;
+
+		dev->lastTxDesc = (dev->lastTxDesc + 1) % SPW_TX_DESC_CNT;
+	}
+
+	TRACE("Packets set up");
+
+	(void)mutexUnlock(dev->txLock);
+
+	return nPackets;
+}
+
+
+static int spw_transmit(spw_dev_t *dev, const uint8_t *buf, const size_t bufsz, const size_t nPackets, bool async)
+{
+	if ((buf == NULL) || (bufsz < SPW_TX_MIN_BUFSZ)) {
+		return -EINVAL;
+	}
+
+	if (nPackets == 0) {
+		return 0;
+	}
+
+	return async ? spw_transmitAsync(dev, buf, nPackets) : spw_transmitWait(dev, buf, nPackets);
+}
+
+
+/* Configure RX DMA descriptors */
+static int spw_rxConfigure(spw_dev_t *dev, size_t *firstDesc, const size_t nPackets)
+{
+	if (nPackets > SPW_RX_DESC_CNT) {
+		return -EINVAL;
+	}
+
+	(void)mutexLock2(dev->rxConfLock, dev->rxLock);
+
+	TRACE("nPackets: %d", nPackets);
+
+	for (size_t cnt = 0; cnt < nPackets; cnt++) {
+		while (!dev->rxAcknowledged[dev->nextRxDesc]) {
+			(void)condWait(dev->rxAckCond, dev->rxLock, 0);
+		}
+
+		dev->rxAcknowledged[dev->nextRxDesc] = false;
+
+		volatile spw_rxDesc_t *desc = &dev->rxDesc[dev->nextRxDesc];
+
+		/* Interrupt after each packet received */
+		desc->ctrl = RX_DESC_IE;
+		if (dev->nextRxDesc == SPW_RX_DESC_CNT - 1) {
+			/* Wrap around */
+			desc->ctrl |= RX_DESC_WR;
+		}
+
+		memset((void *)dev->rxBuff[dev->nextRxDesc], 0, sizeof(dev->rxBuff[dev->nextRxDesc]));
+		desc->addr = va2pa((void *)dev->rxBuff[dev->nextRxDesc]);
+
+		/* Everything is set up, enable descriptor */
+		desc->ctrl |= RX_DESC_EN;
+
+		dev->nextRxDesc = (dev->nextRxDesc + 1) % SPW_RX_DESC_CNT;
+	}
+
+	*firstDesc = (dev->nextRxDesc + SPW_RX_DESC_CNT - nPackets) % SPW_RX_DESC_CNT;
+
+	/* Enable receiver */
+	dev->vbase[DMA_CTRL] |= (DMA_CTRL_RE | DMA_CTRL_RD);
+
+	(void)mutexUnlock(dev->rxLock);
+	(void)mutexUnlock(dev->rxConfLock);
+
+	return nPackets;
+}
+
+
+/* Read from RX buffers */
+static int spw_rxRead(spw_dev_t *dev, size_t firstDesc, uint8_t *buf, size_t bufsz, const size_t nPackets)
+{
+	if (nPackets > SPW_RX_DESC_CNT) {
+		return -EINVAL;
+	}
+
+	size_t cnt = 0;
+	const size_t lastDesc = (firstDesc + nPackets) % SPW_RX_DESC_CNT;
+	bool wrapped = (lastDesc <= firstDesc);
+
+	TRACE("first: %u last: %u nPackets: %u", firstDesc, lastDesc, nPackets);
+
+	(void)mutexLock(dev->rxLock);
+
+	while ((firstDesc < lastDesc) || (wrapped && (firstDesc >= lastDesc))) {
```

Comment:
```suggestion
	while ((firstDesc < lastDesc) || wrapped) {
```
`firstDesc >= lastDesc` is always `true`.